proposals/0303-swiftpm-extensible-build-tools.md

```diff
@@ -287,130 +267,232 @@ protocol TargetBuildContext {
-        inputPaths: [FilePath] = [],
-        outputPaths: [FilePath] = []
+        /// Input files to the build command. Any changes to the input files
+        /// cause the command to be rerun.
+        inputFiles: [Path] = [],
+        /// Output files that should be processed further.
+        outputFiles: [Path] = []
     )
 
-    /// Registers a generated source file that will be passed to later stages
-    /// of the build. Note that this is separate from the output dependencies
-    /// specified when adding a command, which identify the paths of files and
-    /// directories created by the command but which are not treated as source
-    /// files for further processing.
-    func addGeneratedSourceFile(path: FilePath)
-    
-    /// Registers a directory into which a `prebuild` command will write output
-    /// files that should be considered as generated source files of the target
-    /// being built.
+    /// Creates a command to run before the build. The executable should be a
+    /// tool returned by `TargetBuildContext.tool(named:)`, and any paths in
+    /// the arguments list as well as in the input and output lists should be
+    /// based on the paths provided in the target build context structure.
+    ///
+    /// The build command will run before the build starts, and is allows to
+    /// create an arbitrary set of output files based on the contents of the
+    /// inputs.
+    ///
+    /// Because prebuild commands are run on every build, they are can have
+    /// significant performance impact and should only be used when there is
+    /// no way to know the names of the outputs before the command is run.
     ///
-    /// It is an error to call this function from another kind of plugin than
-    /// one that has the `prebuild` capability.
-    func addPrebuildGeneratedSourceDirectory(path: FilePath)
+    /// The `outputDirectory` parameter is the path of a directory into which
+    /// the command will write its output files. Any file in that directory
+    /// will be interpreted according to same build rules as for sources.
+    func addPrebuildCommand(
+        /// An arbitrary string to show in build logs and other status areas.
+        displayName: String,
+        /// The executable to be invoked; should be a tool looked up using
+        /// `tool(named:)`, which may reference either a binary-provided tool
+        /// or a source-built tool.
+        executable: Path,
+        /// Arguments to be passed to the tool. Any paths should be based on
+        /// the paths provided in the target build context.
+        arguments: [String],
+        /// Optional initial working directory of the command.
+        workingDirectory: Path? = nil,
+        /// Any custom settings for the process environment.
+        environment: [String: String] = [:],
+        /// A directory into which the command can write output files that
+        /// should be processed further.
+        outputFilesDirectory: Path
+    )
```

Comment:
oh okey I see, so rather than a prebuild capability we have a command type that we can add here... okey I guess 👍

Reply:
Right, part of the feedback and suggestion from the core team was to combine `prebuild` and `buildTool` into one capability and to use different command constructor methods.  This makes a lot of sense to me.